docs/writing_plugins.md

```diff
@@ -0,0 +1,550 @@
+
+### Exporting
+
+Working with exports is accomplished via the `export` function on the generated file object.  Let's walk through an example:
+
+Suppose you generate a validation function for every message.  If you have a nested message, such as:
+
+```proto
+message Bar {
+   Foo foo = 1;
+}
+```
+
+You may want to import and use the validation function generated for message `Foo` when generating the code for message `Bar`.  To generate the validation function, you would use `export` as follows:
+
+```ts
+const fn = f.export("validateFoo");
+f.print("function ", fn, "() {");
+f.print("   return true;");
+f.print("}");
+```
+
+Note that `export` returns an `ImportSymbol` that can then be used by another dependency.  The trick is to store this `ImportSymbol` and use it when you generate the validation function for `Bar`.  Storing the symbol is as simple as putting it in a global map:
+
+```ts
+const exportMap = new Map<DescMessage, ImportSymbol>()
+```
+
+That way, when you need to use it for `Bar`, you can simply access the map:
+
+```ts
+const fooValidationFn = exportMap.get(bar);  // bar is of type DescMessage
+```
+
+### Parsing plugin options
+
+The plugin framework recognizes a set of pre-defined key/value pairs that can be passed to all plugins when executed (i.e. `target`, `keep_empty_files`, etc.), but if your plugin needs to be passed additional parameters, you can specify a `parseOption` function as part of your plugin initialization.  
+
+```ts
+parseOption(key: string, value: string | undefined): void;
+```
+
+This function will be invoked by the framework, passing in any key/value pairs that it does not recognize from its pre-defined list.
+
+### Reading custom options
+
+Protobuf-ES does not yet provide support for extensions, neither in general as pertaining to proto2 nor with custom options in proto3.  However, in the interim, there are convenience functions for retrieving any custom options specified in your .proto files.  These are provided as a temporary utility until full extension support is implemented.  There are three functions depending on the structure of the custom option desired (scalar, message, or enum):
+
+#### Scalar Options
+
+Custom options of a scalar type (`boolean`, `string`, `int32`, etc.) can be retrieved via the `findCustomScalarOption` function.  It returns a type corresponding to the given `scalarType` parameter.  For example, if `ScalarType.STRING` is passed, the return type will be a `string`.  If the option is not found, it returns `undefined`.
+
+```ts
+function findCustomScalarOption<T extends ScalarType>(
+  desc: AnyDesc,
+  id: number,
+  scalarType: T
+): ScalarValue<T> | undefined;
+```
+
+`AnyDesc` represents any of the `DescXXX` objects such as `DescFile`, `DescEnum`, `DescMessage`, etc.  The `id` parameter represents the ID of the custom options field definition.
```

Comment:
"the ID" doesn't ring right 🤔  

Can we use "extension numbers" ([ref](https://developers.google.com/protocol-buffers/docs/proto#choosing)) ?